source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,333 @@
+
+  DNSServiceErrorType error = pending_resolution->dnsServiceGetAddrInfo(dns_lookup_family);
+  if (error != kDNSServiceErr_NoError) {
+    ENVOY_LOG(error, "DNS resolver error in dnsServiceGetAddrInfo for {}", dns_name);
+    return nullptr;
+  }
+
+  // If the query was synchronously resolved, there is no need to return the query.
+  if (pending_resolution->synchronously_completed_) {
+    return nullptr;
+  }
+
+  pending_resolution->owned_ = true;
+  return pending_resolution.release();
+}
+
+void AppleDnsResolverImpl::addPendingQuery(PendingResolution* query) {
+  ASSERT(queries_with_pending_cb_.count(query) == 0);
+  queries_with_pending_cb_.insert(query);
+}
+
+void AppleDnsResolverImpl::removePendingQuery(PendingResolution* query) {
+  auto erased = queries_with_pending_cb_.erase(query);
+  ASSERT(erased == 1);
+}
+
+void AppleDnsResolverImpl::flushPendingQueries(const bool with_error) {
+  ENVOY_LOG(debug, "DNS Resolver flushing {} queries", queries_with_pending_cb_.size());
+  for (std::set<PendingResolution*>::iterator it = queries_with_pending_cb_.begin();
+       it != queries_with_pending_cb_.end(); ++it) {
+    auto query = *it;
+    try {
+      ASSERT(query->pending_cb_);
+      query->callback_(query->pending_cb_->status_, std::move(query->pending_cb_->responses_));
+    } catch (const EnvoyException& e) {
+      ENVOY_LOG(critical, "EnvoyException in DNSService callback: {}", e.what());
+      query->dispatcher_.post([s = std::string(e.what())] { throw EnvoyException(s); });
+    } catch (const std::exception& e) {
+      ENVOY_LOG(critical, "std::exception in DNSService callback: {}", e.what());
+      query->dispatcher_.post([s = std::string(e.what())] { throw EnvoyException(s); });
+    } catch (...) {
+      ENVOY_LOG(critical, "Unknown exception in DNSService callback");
+      query->dispatcher_.post([] { throw EnvoyException("unknown"); });
+    }
+
+    if (query->owned_) {
+      ENVOY_LOG(debug, "Resolution for {} completed (async)", query->dns_name_);
+      delete *it;
+    } else {
+      ENVOY_LOG(debug, "Resolution for {} completed (synchronously)", query->dns_name_);
+      query->synchronously_completed_ = true;
+    }
+  }
+
+  // Purge the contents so no one tries to delete them again.
+  queries_with_pending_cb_.clear();
+
+  if (with_error) {
+    // The main sd ref is destroyed here because a callback with an error is good indication that
+    // the connection to the DNS server is faulty and needs to be torn down.
+    //
+    // Deallocation of the MainSdRef __has__ to happen __after__ flushing queries. Flushing queries
+    // de-allocates individual refs, so deallocating the main ref ahead would cause deallocation of
+    // invalid individual refs per dns_sd.h
+    ENVOY_LOG(error, "[Error path] deallocating main sd ref");
+    deallocateMainSdRef();
+    ENVOY_LOG(error, "[Error path] re-initializing main sd ref");
+    initializeMainSdRef();
+    ENVOY_LOG(error, "[Error path] done");
```

Comment:
This logs seem marginally useful. I would move to debug/trace and possibly simplify.